salt/crypt.py

```diff
@@ -443,60 +375,111 @@ class MasterKeys(dict):
+        # exist with the same name (master.pem); but with a different driver
+        # the state is shared across all masters, so it would be impossible to
+        # represent that setup unless the key used is unique (e.g the master
+        # id).
+        # when get_keys(name='master') runs it will duplicate the keys to
+        # ${id}.pem/pub to avoid this scenario. at some point in the future
+        # master.pem/pub can be removed
+        self.master_id = self.opts["id"].removesuffix("_master")
+
+        # set names for the signing key-pairs
+        self.pubkey_signature = None
+        self.master_pubkey_signature = (
+            opts.get("master_pubkey_signature") or f"{opts['id']}_pubkey_signature"
+        )
+
+        if autocreate:
+            self._setup_keys()
+
+    # We need __setstate__ and __getstate__ to avoid pickling errors since
+    # some of the member variables correspond to Cython objects which are
+    # not picklable.
+    # These methods are only used when pickling so will not be used on
+    # non-Windows platforms.
+    def __setstate__(self, state):
+        self.__init__(state["opts"])
+
+    def __getstate__(self):
+        return {"opts": self.opts}
+
+    def _setup_keys(self):
+        # it's important to init this even if cluster_id is enabled so that on
+        # initial start the master's non cluster key is generated
         key_pass = salt.utils.sdb.sdb_get(self.opts["key_pass"], self.opts)
-        self.master_key = self.__get_keys(passphrase=key_pass)
 
-        self.cluster_pub_path = None
-        self.cluster_rsa_path = None
-        self.cluster_key = None
-        if self.opts["cluster_id"]:
-            self.cluster_pub_path = os.path.join(
-                self.opts["cluster_pki_dir"], "cluster.pub"
+        if self.cache.contains("master_keys", f"{self.master_id}.pem"):
+            self.master_key = self.key = self.find_or_create_keys(
+                name=self.master_id, passphrase=key_pass
             )
-            self.cluster_rsa_path = os.path.join(
-                self.opts["cluster_pki_dir"], "cluster.pem"
-            )
-            self.cluster_shared_path = os.path.join(
-                self.opts["cluster_pki_dir"],
-                "peers",
-                f"{self.opts['id']}.pub",
+        else:
+            self.master_key = self.key = self.find_or_create_keys(
+                name="master", passphrase=key_pass
             )
+
+            # facilitate migrating to pem named off the master id instead of master.pem
+            if not self.cache.contains("master_keys", f"{self.master_id}.pem"):
+                priv = self.cache.fetch("master_keys", "master.pem")
+                pub = self.cache.fetch("master_keys", "master.pub")
+                self.cache.store("master_keys", f"{self.master_id}.pem", priv)
+                self.cache.store("master_keys", f"{self.master_id}.pub", pub)
+                self.cache.flush("master_keys", "master.pem")
```

Comment:
this is the only notable 'not backwards compatible' thing; i needed to make the master 'rename' master.pub/pem to ${id}.pub/pem, because in a multimaster scenario each masters key needs to be unique, otherwise they would erase each other on startup. i create a symlink (which is ignored in localfs_keys_backcompat) but not on windows because i was seeing some strange behavior / fails in the tests

Reply:
I'd like to have @dwoz look at this to make sure it's good

Reply:
going to re-run CI one more time to see if symlink on windows actually works. still worth calling out this change either way

Reply:
seems to work actually. strange :) ill leave it as is so symlink should be on all platforms